extension-manifest-v3/src/background/adblocker.js

```diff
@@ -286,12 +297,97 @@ chrome.webNavigation.onCommitted.addListener(async (details) => {
   });
 });
 
-chrome.runtime.onMessage.addListener((msg, sender) => {
-  if (msg.action === 'getCosmeticsFilters') {
-    adblockerOnMessage(msg, sender).catch((e) =>
-      console.error(`Error while processing cosmetics filters: ${e}`),
-    );
-  }
+if (__PLATFORM__ === 'firefox') {
+  chrome.webRequest.onBeforeRequest.addListener(
+    (details) => {
+      const isMainFrame = details.type === 'main_frame';
+      const sourceUrl = isMainFrame
+        ? details.url
+        : details.originUrl || details.documentUrl;
 
-  return false;
-});
+      const parsedUrl = parse(details.url);
+      const parsedSourceUrl = isMainFrame ? parsedUrl : parse(sourceUrl);
+
+      const request = new Request({
+        requestId: details.requestId,
+        tabId: details.tabId,
+
+        domain: parsedUrl.domain,
+        hostname: parsedUrl.hostname,
+        url: details.url.toLowerCase(),
+
+        sourceDomain: parsedSourceUrl.domain || '',
+        sourceHostname: parsedSourceUrl.hostname || '',
+        sourceUrl: sourceUrl.toLowerCase(),
```

Comment:
sourceUrl can be "undefined" (details.originUrl and details.documentUrl being absent). A test page is to play a YouTube video.

Reply:
I will add `|| ''` to the `sourceUrl` definition.

Reply:
https://github.com/ghostery/ghostery-extension/pull/1048/files#diff-df2e21bb6266c3df460f6af7886b15496d496b3bd46182750a55addf0a1a7a89R306